### ISSUE TITLE
Define use cases

### ISSUE THREAD

**Issue by grezniczek** [davwillev/stopwatch#2]
@davwillev - We should first define the concrete use cases the stopwatch EM is supposed to cover.
There are probably VERY many, such as:

- Simple starting/stoping and putting the elapsed time into a text field.
  - This could be in ms (integer),
  - or seconds (decimal number),
  - or minutes,
  - or formatted HH:mm:ss.nnn
- Capture start time, stop time, and elapsed time in 3 different fields
  - e.g. 2x datetime w/seconds, and an integer field for seconds (or decimal when including fractional seconds).
  - (unvalidated) text fields with HH:mm.ss.nnn format,
  - maybe include the date as well (optionally)
- Capture start, stop, lap time points
  - with or without date (formatted a specific way),
  - with or without the option to halt the time (without reseting - as in current code),
  - as JSON in a text box,
  - as text in a notes box,
  - as new instances in a separate repeating instrument.

This get's rather complicated very quickly. All needs to be configureable by the EM user aka project designer. Fields need to be mapped, etc....

Maybe start with just one use case, but keeping the design open to accomodate more/all of them.

**Comment by davwillev** [comment]
Thanks for this, @grezniczek  
I went through a similar thought process when first thinking through what the SW EM should look like and how a user should interact with it.

Our initial use case was the '[Timed Up and Go](https://www.cdc.gov/steadi/pdf/TUG_Test-print.pdf)' test, which was typically being performed by a researcher using a separate stopwatch and the result manually being written into a REDCap text field. This requires simple start/stop/capture functionality but was prone to user variations in formatting and copying errors. Hence, direct time capture is desirable as it would remove these issues and the need for separate hardware (i.e. stopwatch).

Once one starts to think through functionality of a stopwatch field, laps/splits quickly come to mind of course. However, on an actual stopwatch each lap typically starts counting from zero again after the previous lap is captured; whereas, it is probably more useful for us to capture time instances from start (unless the counter is paused) and calculate relative differences afterwards, if desired. Allowing for continuous captures also provides timestamped counts over a period of time ('time-tallying') which allows one to calculate a rate of observed events. There are several physical tests that require a count of events over a (possibly arbitrary) period of time, e.g. balance tests, behavioural assessments, and even heart rate could be calculated in this way (and I'm sure you can think of some more examples).

For the UI, I think we should format in HH:mm:ss.nnn. Days is probably overkill. Furthermore, if users are going to perform calculations using their stopwatch results, or use these for branching logic, we should probably capture results in ms in a simple text box (validated in REDCap as an integer). We have used an array to capture time instances in stopwatch.js, which means we don't have to specify the number of captures in advance, but what is the best way to save each as a new REDCap variable? Andy Martin suggested using JSON and passing this to a repeating instrument; I'm not sure of our other options here.

I agree that it would be useful to capture current datetime upon start/stop/capture, but I'm not sure how we can mary our 1/1000s accuracy time with the 1s accuracy of the current REDCap datetime. If datediff was able to calulate even 1/100 second accuracy, we probably wouldn't need to build a SW EM, but unfortunately it doesn't. I agree that we should keep our design open in case users want the datetime functionality; indeed, the stopwatch.js file does capture the current date and time in ms so we could probably add this as an option fairly easily.

**Comment by grezniczek** [comment]
Let's settle on Stopwatch EM **requiring** a text field (optionally with formatting) as a target field.
When a format is set, Stopwatch will perform accordingly (unless set otherwise by action tag parameters, which are still to be determined).
We need at least one parameter for the action tag: the name of the field to store the value in. This will allow the action tag to be on e.g. a descriptive field, and the capture field could be hidden  and/or read only. Since there will be more configuration in the future, the parameter should be valid JSON. E.g. `{"elapsed": "name_of_target_field"}`. The target field will still need to be on the same form (or section in case of a paged survey) as the action tag referring to it.

Behaviour, based on field type, will be this:
- integer: Milliseconds elapsed between starting and stopping (seconds can easily be deduced from this in a calculated field)
- number / number with comma: Seconds (with fractional seconds) elapsed.
- MM:SS: Just that. Note that this is limited to 59:59.
To be implemented later:
- none: full featured, all data stored as JSON.

Furthermore: additional parameters for the action tag, like specifying additional target fields for start / end time.

This setup will not require any configuration logic. It's all in the action tag.
For advanced scenarios, e.g. creating repeat instances of another form, configuration via the module config may be a better way than to have complex JSON in the action tag.

What do you think?
This should cover the Timed Up and Go test. Incidentally, colleagues of mine are just about to start a new project where this is assessed. They might like this EM, too.

**Comment by davwillev** [comment]
That all sounds very good and sensible.

May I ask, how do you foresee the implementation of multiple captures (laps) into REDCap variables? Would this be something we have to look at after the basic implemntation?

**Comment by grezniczek** [comment]
@davwillev, yes. I guess the repeating instance approach suggested by Andy Martin will be the way to go.

And yes, I think this will be the second step after getting the basics going.
I added a new [development](https://github.com/davwillev/stopwatch/tree/development) branch that as of this writing (https://github.com/davwillev/stopwatch/commit/7d9bf6c25019713660e3033280e4adead5db6f49) is a "functioning" EM that can be installed and activated in projects and fields (with `@STOPWATCH`). All it does for now is output "Stopwatch" in the fields tagged with the action tag.

**Comment by davwillev** [comment]
That's great @grezniczek 
Run with it as far as you wish; it's further than we could get :)

**Comment by grezniczek** [comment]
I'll try to get the very basic start/stop/capture elapsed functionality working today.
